light-client/src/builder/light_client.rs

```diff
@@ -0,0 +1,213 @@
+//! DSL for building a light client [`Instance`]
+
+use tendermint::{block::Height, Hash};
+
+use crate::bail;
+use crate::builder::error::{self, Error};
+use crate::components::clock::Clock;
+use crate::components::io::{AtHeight, Io};
+use crate::components::scheduler::Scheduler;
+use crate::components::verifier::Verifier;
+use crate::light_client::{LightClient, Options};
+use crate::operations::Hasher;
+use crate::predicates::VerificationPredicates;
+use crate::state::{State, VerificationTrace};
+use crate::store::LightStore;
+use crate::supervisor::Instance;
+use crate::types::{LightBlock, PeerId, Status};
+
+#[cfg(feature = "rpc-client")]
+use {
+    crate::components::clock::SystemClock, crate::components::io::ProdIo,
+    crate::components::scheduler, crate::components::verifier::ProdVerifier,
+    crate::operations::ProdHasher, crate::predicates::ProdPredicates, std::time::Duration,
+    tendermint_rpc as rpc,
+};
+
+/// No trusted state has been set yet
+pub struct NoTrustedState;
+
+/// A trusted state has been set and validated
+pub struct HasTrustedState;
+
+/// Builder for a light client [`Instance`]
+#[must_use]
+pub struct LightClientBuilder<State> {
+    peer_id: PeerId,
+    options: Options,
+    io: Box<dyn Io>,
+    clock: Box<dyn Clock>,
+    hasher: Box<dyn Hasher>,
+    verifier: Box<dyn Verifier>,
+    scheduler: Box<dyn Scheduler>,
+    predicates: Box<dyn VerificationPredicates>,
+    light_store: Box<dyn LightStore>,
+
+    #[allow(dead_code)]
+    state: State,
```

Comment:
We need a field here because in Rust type parameters of a struct have to appear in its fields.

We could use `PhantomData<State>`, but since the two type states are zero-sized anyway (same as `PhantomData`), this saves us an import here and some potential confusion wrt to `PhantomData`.